### PR TITLE
Handle missing/invalid session state gracefully

### PR DESCRIPTION
**Pull request by janakamarasena**
### Proposed changes in this pull request

fixes wso2/product-is#4118

Adding the following config in the identity.xml file can disable this behaviour
`<HandleAlreadyLoggedOutSessionsGracefully>false</HandleAlreadyLoggedOutSessionsGracefully>
`

### Follow up actions
Need to mention in 5.8.0 migration docs as default behavior has changed. Default behavior can be taken by adding the `<HandleAlreadyLoggedOutSessionsGracefully>false</HandleAlreadyLoggedOutSessionsGracefully>
`
config to the identity.xml file.

Changed behaviour: In a logout request if the opbs cookie or a valid session does not exist instead of showing the invalid request error page the user will be redirected to the successfully logged out page of the IS or if a valid id_token_hint and a valid post_logout_redirect_uri is available user will be redirected to the post_logout_redirect_uri.

<br><br><br><br><br><br>
### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

**Comment by mefarazath** [comment]
@janakamarasena create an issue in product-is with the migration requirement. Need to add "migration" label as well.

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java [hunk 1]
```diff
@@ -639,4 +659,42 @@ private String getServiceProviderName(String clientId, String tenantDomain) {
...
+        String redirectURL = OIDCSessionManagementUtil.getOIDCLogoutURL();
+        String idTokenHint = request.getParameter(OIDCSessionConstants.OIDC_ID_TOKEN_HINT_PARAM);
+        String postLogoutRedirectUri = request.getParameter(OIDCSessionConstants.OIDC_POST_LOGOUT_REDIRECT_URI_PARAM);
+        if (StringUtils.isEmpty(idTokenHint) || StringUtils.isEmpty(postLogoutRedirectUri)) {
+            response.sendRedirect(getRedirectURL(redirectURL, request));
+            return;
+        }
+        if (!validateIdToken(idTokenHint)) {
+            String msg = "ID token signature validation failed.";
+            log.error(msg);
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java]: This error log seems not useful. Can we add some context information to this to identify which validation failed? If we can't we may not need the log, as we already respond with the same error

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java [hunk 2]
```diff
@@ -639,4 +659,42 @@ private String getServiceProviderName(String clientId, String tenantDomain) {
...
+            } else {
+                redirectURL = OIDCSessionManagementUtil.getErrorPageURL(OAuth2ErrorCodes.ACCESS_DENIED,
+                        "Post logout URI does not match with registered callback URI.");
+            }
+        } catch (ParseException e) {
+            String msg = "Error occurred while getting application information.";
+            log.error(msg, e);
+            redirectURL = OIDCSessionManagementUtil.getErrorPageURL(OAuth2ErrorCodes.ACCESS_DENIED, msg);
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            String msg = "Error occurred while getting application information. Client id not found.";
```
Comment (pulasthi7) [inline components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java]: Add context information to these logs, to identify which request/client caused the issue

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1130

File: components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java (hunk 1 above)
Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1130

**Pull request merged**